src/gpl/src/initialPlace.cpp

```diff
@@ -106,318 +96,356 @@ void InitialPlace::doBicgstabPlace() {
 
   // set ExtId for idx reference // easy recovery
   setPlaceInstExtId();
-  for(int i=1; i<=ipVars_.maxIter; i++) {
+
+  for (size_t iter = 1; iter <= ipVars_.maxIter; iter++) {
     updatePinInfo();
     createSparseMatrix();
+#ifdef ENABLE_GPU
+  if (!ipVars_.forceCPU){
+    int gpu_count = 0;
+    cudaGetDeviceCount(&gpu_count);	
+    if(gpu_count != 0){
+      // CUSOLVER based on sparse matrix and QR decomposition for initial place
+      error = cudaSparseSolve(iter, placeInstForceMatrixX_,
+                      fixedInstForceVecX_,
+                      instLocVecX_,
+                      placeInstForceMatrixY_,
+                      fixedInstForceVecY_,
+                      instLocVecY_,
+                      log_);
+    }
+    else{
+      log_->warn(GPL, 250, "GPU is not available. CPU solver is automatically used.");
+      // BiCGSTAB solver for initial place
+      error = cpuSparseSolve(ipVars_.maxSolverIter, iter,
```

Comment:
you have this same block of code three times - refactor

Reply:
I think the three blocks are necessary. The program gives different warnings under different circumstances. Do you want me to wrap it and write it as a shorter sentence?

Reply:
Just use a simple boolean and an if statement (if (run_cpu) ....)

Reply:
Get it. Thanks.